packages/babel-core/src/config/full.ts

```diff
@@ -488,16 +488,18 @@ function* loadPresetDescriptor(
   };
 }
 
-function chain<Args extends any[]>(
-  a: undefined | ((...args: Args) => void),
-  b: undefined | ((...args: Args) => void),
-) {
-  const fns = [a, b].filter(Boolean);
-  if (fns.length <= 1) return fns[0];
-
-  return function (this: unknown, ...args: unknown[]) {
-    for (const fn of fns) {
-      fn.apply(this, args);
+function chainMaybeAsync<Args extends any[], R extends void | Promise<void>>(
+  a: undefined | ((...args: Args) => R),
+  b: undefined | ((...args: Args) => R),
```

Comment:
Here we assume `a` and `b` share the same return type, however, it is also possible that `a` is sync while `b` is async. I suggest we remove the `R` constraint and type them as `(...args: Args) => void | Promise<void>`. We can probably get rid of the final type cast as well.

For situations where both `a` and `b` are sync, we can add a type overload so that the chained function is still sync.

Reply:
On the mixed-type cases, for example, if `inherits.pre` is a sync function and `plugin.pre` is an async function, then the merged `pre` will be an async function.

Like you said, every function is assignable to `() => void`, so the type checking can't help very much here. I am fine with leaving it as-is.